runtime/sema/check_assignment.go

```diff
@@ -107,6 +110,76 @@ func (checker *Checker) checkAssignment(
 	return
 }
 
+func (checker *Checker) enforcePureAssignment(assignment ast.Statement, target ast.Expression) {
+	if !checker.CurrentPurityScope().EnforcePurity {
+		return
+	}
+
+	var variable *Variable
+
+	switch targetExp := target.(type) {
+	case *ast.IdentifierExpression:
+		variable = checker.valueActivations.Find(targetExp.Identifier.Identifier)
+	case *ast.IndexExpression:
+		if indexIdentifier, ok := targetExp.TargetExpression.(*ast.IdentifierExpression); ok {
+			variable = checker.valueActivations.Find(indexIdentifier.Identifier.Identifier)
+		}
+	case *ast.MemberExpression:
+		if indexIdentifier, ok := targetExp.Expression.(*ast.IdentifierExpression); ok {
+			variable = checker.valueActivations.Find(indexIdentifier.Identifier.Identifier)
+		}
+	}
+
+	// if the target is not a variable (e.g. a nested index expression x[0][0], or a nested
+	// member expression x.y.z), the analysis cannot know for sure where the value being
+	// assigned to originated, so we must default to impure. Consider:
+	// -----------------------------------------------------------------------------
+	// let a: &[Int] = &[0]
+	// pure fun foo(_ x: Int) {
+	//     let b: &[Int] = [0]
+	//     let c = [a, b]
+	//     c[x][0] = 4 // we cannot know statically whether a or b receives the write here
+	// }
+	if variable == nil {
+		checker.ObserveImpureOperation(assignment)
+		return
+	}
+
+	// `self` technically exists in param scope, but should still not be writeable
+	// outside of an initializer
+	if variable.DeclarationKind == common.DeclarationKindSelf {
+		if checker.functionActivations.Current().InitializationInfo == nil {
+			checker.ObserveImpureOperation(assignment)
+		}
+		return
+	}
+
+	// We have to prevent any writes to references, since we cannot know where the value
+	// pointed to by the reference may have come from. Similarly, we can never safely assign
+	// to a resource; because resources are moved instead of copied, we cannot currently
+	// track the origin of a write target when it is a resource. Consider:
+	// -----------------------------------------------------------------------------
+	// pub resource R {
+	//   pub(set) var x: Int
+	//   init(x: Int) {
+	//     self.x = x
+	//   }
+	//  }
+	//
+	// pure fun foo(_ f: @R): @R {
+	//   let b <- f
+	//   b.x = 3 // b was created in the current scope but modifies the resource value
+	//   return <-b
+	// }
+	if _, ok := variable.Type.(*ReferenceType); ok || variable.Type.IsResourceType() ||
+		// when the variable is neither a resource nor a reference, we can write to if its
+		// activation depth is greater than or equal to the depth at which the current purity
+		// scope was created; i.e. if it is a parameter to the current function or was created
+		// within it
+		checker.CurrentPurityScope().ActivationDepth > variable.ActivationDepth {
```

Comment:
Maybe improve indentation, the last case (activation depth) looks like part of the body:
```suggestion
	//
	// When the variable is neither a resource nor a reference, 
	// we can write to if its activation depth is greater than or equal to the depth 
	// at which the current purity scope was created; 
	// i.e. if it is a parameter to the current function or was created within it.
	//
	if _, ok := variable.Type.(*ReferenceType); ok || 
		variable.Type.IsResourceType() ||
		checker.CurrentPurityScope().ActivationDepth > variable.ActivationDepth {

```